### ISSUE TITLE
Hamming distance implementation

### ISSUE THREAD

**Issue by ideasbyjin** [maxbachmann/rapidfuzz-cpp#20]
Hi!

I love this library and how it provides such a rapid distance calculation for Python, so I've created a Hamming distance implementation. However I can't seem to create a pull request from my branch. Would it be possible to merge this, as I'd love to have more distance metrics being available as part of this library.

```c++
// levenshtein.hpp
template <typename Sentence1, typename Sentence2>
std::size_t hamming(const Sentence1& s1, const Sentence2& s2);

// levenshtein.txx
template <typename Sentence1, typename Sentence2>
std::size_t levenshtein::hamming(const Sentence1& s1, const Sentence2& s2)
{
  auto sentence1 = utils::to_string_view(s1);
  auto sentence2 = utils::to_string_view(s2);

  // Swapping the strings so the first string is shorter
  if (sentence1.size() != sentence2.size()) {
    return std::numeric_limits<std::size_t>max();
  }

  std::size_t hamm = 0;

  for (std::size_t i = 0; i < sentence1.length(); i++) {
      if (sentence1[i] != sentence2[i]) {
          hamm++;
      }
  }
  return hamm;
}
```

**Comment by ideasbyjin** [comment]
For what it's worth, I've also updated the `py_levenshtein.cpp` code in the Python repo:
```
constexpr const char* hamming_docstring = R"(
Calculates the Hamming distance between two strings.

Args:
    s1 (str):  first string to compare
    s2 (str):  second string to compare

Returns:
    int: Hamming distance between s1 and s2
)";

// C++11 does not support generic lambdas
struct HammingDistanceVisitor {
  template <typename Sentence1, typename Sentence2>
  std::size_t operator()(Sentence1&& s1, Sentence2&& s2) const
  {
    return rlevenshtein::hamming(s1, s2);
  }
};

PyObject* hamming(PyObject* /*self*/, PyObject* args, PyObject* keywds)
{
  PyObject* py_s1;
  PyObject* py_s2;
  static const char* kwlist[] = {"s1", "s2", NULL};

  if (!PyArg_ParseTupleAndKeywords(args, keywds, "OO", const_cast<char**>(kwlist), &py_s1, &py_s2))
  {
    return NULL;
  }

  if (!valid_str(py_s1, "s1") || !valid_str(py_s2, "s2")) {
    return NULL;
  }

  auto s1_view = decode_python_string(py_s1);
  auto s2_view = decode_python_string(py_s2);
  std::size_t result = mpark::visit(HammingDistanceVisitor(), s1_view, s2_view);

  return PyLong_FromSize_t(result);
}
```

**Comment by ideasbyjin** [comment]
@maxbachmann the way I'm catching the length mismatch isn't quite right but would love your thoughts!

**Comment by maxbachmann** [comment]
Thank you for the contribution @ideasbyjin.

I would change the length mismatch error to:
```cpp
  if (sentence1.size() != sentence2.size()) {
    throw std::invalid_argument("s1 and s2 are not the same length.");
  }
```

```cpp
std::size_t result;
try {
    result = mpark::visit(HammingDistanceVisitor(), s1_view, s2_view);
} catch(std::invalid_argument& e) {
    PyErr_Format(PyExc_ValueError, e.what());
    return NULL;
}
```

and update the docstring to tell the user, that the lenght has to be similar (otherwise a ValueError is thrown)

**Comment by maxbachmann** [comment]
I added the implementation. The commits reference you as co author.

**Comment by ideasbyjin** [comment]
@maxbachmann many thanks! This is great. I realised the Python library (`py_levenshtein.cpp`) still remains unchanged though, can we get that sorted?

**Comment by maxbachmann** [comment]
I added this in the Python library yesterday aswell: https://github.com/maxbachmann/RapidFuzz/commit/fa17ff98e8dba3d67e21b4264d8eca336a8e401a

I released Version 0.14.0 yesterday, that includes this change.

**Comment by ideasbyjin** [comment]
doh thanks i'm blind
